Makefile

```diff
@@ -742,6 +742,9 @@ $(BUILD_DIR)/baserom/%.o: $(EXTRACTED_DIR)/baserom/%
 $(BUILD_DIR)/data/%.o: data/%.s
 	$(CPP) $(CPPFLAGS) -Iinclude $< | $(AS) $(ASFLAGS) -o $@
 
+$(BUILD_DIR)/assets/text/%.enc.chn.h: assets/text/%.h $(EXTRACTED_DIR)/text/%.h assets/text/charmap.txt
+	$(CPP) $(CPPFLAGS) -I$(EXTRACTED_DIR) $< | $(PYTHON) tools/msgenc.py --encoding chn --charmap assets/text/charmap.txt - $@
+
```

Comment:
shouldn't we use NES everywhere instead of a separate CHN for all of these (as in, DEFINE_MESSAGE_NES instead of DEFINE_MESSAGE_CHN)? Especially since we'd need different charmaps for the two Chinese versions anyway. I think that's how they did it originally and it would simplify a lot of things

Reply:
I'd like the macro to still be `DEFINE_MESSAGE(jpn, nes, ger, fra, chn)` so that it's easier for modifications to merge them.

Reply:
Hmm, how does it make merging easier? If you wanted text from multiple game versions in the same text file I think I would have to write a script anyway, in which case having the `chn` argument doesn't really make things easier for me.

One nice thing about `DEFINE_MESSAGE(jpn, nes, fra, ger)` is that the four message types correspond to the four message_data_static files and message entry tables. I think we should stop treating these as "languages", since NES is not necessarily english and we don't know where the name comes from anyway. Instead, we could have two different concepts:

* "Message sets" jpn/nes/fra/ger/staff, used everywhere in the game code. At runtime, the only thing that really matters is whether these use the "wide" (kanji) or "non-wide" (nes_font_static) encoding. 
* "Languages" jp/en/fr/de/zh-Hans/zh-Hant, which say how to encode/decode messages with Unicode for assets. I'm not sure how to name these, but I don't think we need to use 3 characters since they're a separate thing from jpn/nes/fra/ger. Maybe "sc" and "tc" for chinese? (I think "ique-cn" is probably a bad name since I hear Traditional Chinese is also used in Hong Kong and Macau, and Taiwan is still called "China", etc.) I'm also not sure if en/fr/de are the same language ("latin"?) or different languages that have identical charsets.

If we go this route, I think it could look something like this eventually (maybe the charmap knows about variable-width characters, idk):
```make
$(BUILD_DIR)/assets/text/%.enc.jpn.h: assets/text/%.h $(EXTRACTED_DIR)/text/%.h assets/text/charmap.jp.txt
	$(CPP) $(CPPFLAGS) -I$(EXTRACTED_DIR) $< | $(PYTHON) tools/msgenc.py --charmap assets/text/charmap.jp.txt - $@

ifeq ($(VERSION),ique-sc)
$(BUILD_DIR)/assets/text/%.enc.nes.h: assets/text/%.h $(EXTRACTED_DIR)/text/%.h assets/text/charmap.sc.txt
	$(CPP) $(CPPFLAGS) -I$(EXTRACTED_DIR) $< | $(PYTHON) tools/msgenc.py --charmap assets/text/charmap.sc.txt - $@
else ifeq ($(VERSION),ique-tc)
$(BUILD_DIR)/assets/text/%.enc.nes.h: assets/text/%.h $(EXTRACTED_DIR)/text/%.h assets/text/charmap.tc.txt
	$(CPP) $(CPPFLAGS) -I$(EXTRACTED_DIR) $< | $(PYTHON) tools/msgenc.py --charmap assets/text/charmap.tc.txt - $@
else
$(BUILD_DIR)/assets/text/%.enc.nes.h: assets/text/%.h $(EXTRACTED_DIR)/text/%.h assets/text/charmap.en.txt
	$(CPP) $(CPPFLAGS) -I$(EXTRACTED_DIR) $< | $(PYTHON) tools/msgenc.py --charmap assets/text/charmap.en.txt - $@
endif

$(BUILD_DIR)/assets/text/%.enc.ger.h: assets/text/%.h $(EXTRACTED_DIR)/text/%.h assets/text/charmap.de.txt
	$(CPP) $(CPPFLAGS) -I$(EXTRACTED_DIR) $< | $(PYTHON) tools/msgenc.py --charmap assets/text/charmap.de.txt - $@

$(BUILD_DIR)/assets/text/%.enc.fra.h: assets/text/%.h $(EXTRACTED_DIR)/text/%.h assets/text/charmap.fr.txt
	$(CPP) $(CPPFLAGS) -I$(EXTRACTED_DIR) $< | $(PYTHON) tools/msgenc.py --charmap assets/text/charmap.fr.txt - $@
```

What are your thoughts?

Reply:
(your NES_CHARMAP is probably cleaner, I forgot about that when writing this comment)

Reply:
I like the sound of distinguishing "message sets" and "languages". I'm not sure about splitting the charmap for en/fr/de, I feel like that shouldn't be necessary. I'll look into it some more